machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,218 @@
+	// startTime describes when this build began
+	// +kubebuilder:validation:Required
+	StartTime *metav1.Time `json:"startTime"`
+	// endTime describes when the build ended
+	// +optional
+	EndTime *metav1.Time `json:"endTime"`
+	// buildHistory contains previous iterations of failed or interrupted builds related to this one.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	// prepared indicates that the build has finished preparing. A build is prepared
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuilding BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// restarted indicates that this build has been either interrupted or failed, and now a new
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
+	// ready indicates that the build has completed and the image is ready to roll out.
+	MachineOSReady BuildProgress = "Ready"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type BuildHistory struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+	// configMaps references all config map objects created during the build process
+	// +optional
+	ConfigMaps []BuildConfigMapObjectReference `json:"configmaps"`
+	// buildPod references the build pod used in this previous build if it existed
+	// +optional
+	BuildPod BuildPodReference `json:"buildPod"`
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
+type BuildPodReference struct {
+	// name of the pod
+	Name string `json:"name"`
+	// exitReason is how the pod exited
+	ExitReason string `json:"exitReason"`
+}
+
+// BuoldConfigMapObjectReference refers to the name and contents of a configmap used in a previous build
+type BuildConfigMapObjectReference struct {
+	// name of the configmap
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// Data is the string data in the configmap
+	// +optional
+	Data ConfigMapData `json:"data"`
```

Comment:
Configmap data will always been small in this case, sorry I should specify this. @cheesesashimi do you have a rough estimate of how many entries are in these MCO namespaced OCB configmaps?

These are just used to track temp data for the build.

Reply:
To give a bit of context behind what these ConfigMaps are and why they're used:

These ConfigMaps contain the rendered Containerfile and the rendered MachineConfigs as build inputs. The Containerfile is reasonably small, but a MachineConfig might be a bit larger. In some cases, it is not unheard of for a rendered MachineConfig to be greater than 1 MB in size (once serialized to JSON). The reason I know this is because this became a concern for Hypershift, which stuffs rendered MachineConfigs into ConfigMaps and we had to implement gzip / base-64 encoding to work around the 1 MB size limit that ConfigMaps have.

We do a similar thing in the BuildController to avoid giving the build pods elevated permissions. We stuff the rendered Containerfile and rendered MachineConfigs into ConfigMaps and mount them into the build context. These exist for the duration of the build. Assuming the build is successful, these objects are deleted. If the build is unsuccessful, the objects remain so that a cluster admin can troubleshoot.

Doing some back-of-napkin math here, let's assume the worst-case that we have an OpenShift cluster with two MachineConfigPools, one of which has both amd64 and aarch64 nodes:

Control Plane Pool:
- Rendered MachineConfig (1 MB max size)
- Rendered Containerfile (1 MB max size)
2 MB total, worst-case.

Worker Pool:
- Rendered MachineConfig (1 MB max size) - There is no way to set an architecture-specific MachineConfig, so this ConfigMap could be shared between both the amd64 and aarch64 builds.
- Rendered AMD64 Containerfile (1 MB max size)
- Rendered ARM64 Containerfile (1 MB max size) - We could dedupe these if we've determined that the Containerfiles are equal.
3 MB total, worst-case.

Cluster total: 5 MB, worst-case. This is assuming that one is running all of the builds concurrently. Assuming no builds are running, these ConfigMaps will not exist. Each additional pool would be at minimum, 2 MB for homogenous clusters, with 1 MB additional for each additional architecture supported.

I hope this helps to answer the question. My initial thought is that maybe we shouldn't inline this data. But I could be convinced otherwise.

Reply:
For a more realistic example, on a freshly-provisioned OpenShift cluster with on-cluster builds enabled, I pulled down the ephemeral ConfigMaps to get an idea of the size of their contents:

```console
# This is the size of the rendered MachineConfig for this pool once it's been gzipped / base64-encoded, and stuffed into a ConfigMap.
$ oc get configmap/mc-rendered-layered-950b76f1b1fb344d902cb041ad15d695 -n openshift-machine-config-operator -o json | jq -c -r '.data["machineconfig.json.gz"]' | wc -c
   49825

# This is the size of the rendered Containerfile for this pool once it's been rendered and stuffed into a ConfigMap.
$ oc get configmap/dockerfile-rendered-layered-950b76f1b1fb344d902cb041ad15d695 -n openshift-machine-config-operator -o json | jq -c -r '.data.Dockerfile' | wc -c
    3327
```

These numbers are probably much closer to the average case.

Reply:
Thanks for the details my concern with inlining this data into the Spec would be with the size limit of the CR.  I believe you would be safer having references to resources but then you need to maintain them. Would limiting the size of these fields also limit future use cases?